les/lespay/server/balance.go

```diff
@@ -337,69 +101,82 @@ func (n *NodeBalance) GetBalance() (uint64, uint64) {
 	return n.balance.pos.Value(n.bt.posExp.LogOffset(now)), n.balance.neg.Value(n.bt.negExp.LogOffset(now))
 }
 
-// AddPosBalance adds the given amount to the positive balance and returns the balance
+// GetRawBalance returns the current positive and negative balance
+// but in the raw(expired value) format.
+func (n *NodeBalance) GetRawBalance() (utils.ExpiredValue, utils.ExpiredValue) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+
+	now := n.bt.clock.Now()
+	n.updateBalance(now)
+	return n.balance.pos, n.balance.neg
+}
+
+// AddBalance adds the given amount to the positive balance and returns the balance
 // before and after the operation. Exceeding maxBalance results in an error (balance is
 // unchanged) while adding a negative amount higher than the current balance results in
 // zero balance.
-func (n *NodeBalance) AddPosBalance(amount int64) (uint64, uint64, error) {
-	n.bt.lock.Lock()
+func (n *NodeBalance) AddBalance(amount int64) (uint64, uint64, error) {
 	n.lock.Lock()
-	var setPriority bool
-	defer func() {
-		n.lock.Unlock()
-		n.bt.lock.Unlock()
-		if setPriority {
-			n.bt.ns.SetState(n.node, n.bt.PriorityFlag, nodestate.Flags{}, 0)
-		}
-		n.signalPriorityUpdate()
-	}()
-
 	now := n.bt.clock.Now()
 	n.updateBalance(now)
-	logOffset := n.bt.posExp.LogOffset(now)
-	oldValue := n.balance.pos.Value(logOffset)
+
+	// Ensure the given amount is valid to apply.
+	offset := n.bt.posExp.LogOffset(now)
+	oldValue := n.balance.pos.Value(offset)
 	if amount > 0 && (amount > maxBalance || oldValue > maxBalance-uint64(amount)) {
 		return oldValue, oldValue, errBalanceOverflow
 	}
-	n.bt.activeBalance.SubExp(n.balance.pos)
-	n.balance.pos.Add(amount, logOffset)
-	n.bt.activeBalance.AddExp(n.balance.pos)
-	setPriority = n.checkPriorityStatus()
+
+	// Update the total positive balance counter.
+	defer func(origin utils.ExpiredValue, active bool) {
+		n.bt.adjustBalance(origin, n.balance.pos, active)
```

Comment:
This can lead to a data race but I think I have a nice solution (see my comment at `deactivate`).

Reply:
right, fixed